### PR TITLE
basic maintenance mode via rack

### PR DESCRIPTION
**Pull request by samsonasu**

### CODE REVIEW

File: lib/jefferies_tube/capistrano/db.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+  desc "Fetch the latest database backup"
+  task :fetch do
+    on roles(:db), primary: true do |host|
+      FileUtils.mkdir_p 'db/backups'
+      download! "#{deploy_to}/shared/db/backups/latest.dump", "db/backups/latest-#{fetch(:stage)}.dump"
+    end
+  end
+
+  desc "restore the database from a local file FILE=./local/file_path"
```
Comment (boatrite) [inline lib/jefferies_tube/capistrano/db.rb]: "restore" -> "Restore"

File: lib/jefferies_tube/rack/maintenance.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+      attr_reader :app, :options
+
+      def initialize(app, options={})
+        @app     = app
+        @options = options
+      end
+
+      def call(env)
+        #data = File.read(file)
+        data = "Sorry, this site is down for maintenance"
```
Comment (Nonsense331) [inline lib/jefferies_tube/rack/maintenance.rb]: Does this not use the message from the file?

File: lib/jefferies_tube/capistrano/deploy.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
+namespace :deploy do
+  task :ensure_tag do
+    tagname = `git describe --exact-match HEAD`.chomp
+    if $? == 0
+      puts "Code is tagged as `#{tagname}`, proceeding with deployment"
+    else
+      abort "You need to tag the source before you can deploy production"
+    end
+  end
+  task :create_tag do
```
Comment (boatrite) [inline lib/jefferies_tube/capistrano/deploy.rb]: Put lines between tasks

File: lib/jefferies_tube/capistrano/maintenance.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
+namespace :maintenance do
+  desc "Enable maintenance: set MESSAGE='We should be back in approximately 2 hours'"
+  task :enable do
+    on roles(:web) do |host, user|
+      within current_path do
+        message = ENV["MESSAGE"] || "We should return shortly."
```
Comment (boatrite) [inline lib/jefferies_tube/capistrano/maintenance.rb]: Maybe put the word "maintenance" in the default message somewhere. And why not make the default message the same as in the description?

File: lib/jefferies_tube/rack/maintenance.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    class Maintenance
+      attr_reader :app, :options
+
+      def initialize(app, options={})
+        @app     = app
+        @options = options
+      end
+
+      def call(env)
+        #data = File.read(file)
```
Comment (boatrite) [inline lib/jefferies_tube/rack/maintenance.rb]: :dash: ?

**Comment by Nonsense331** [comment]
🚢

File: lib/jefferies_tube/railtie.rb [hunk 6]
```diff
@@ -18,7 +18,22 @@ class Railtie < ::Rails::Railtie
...
+    end
+
+    initializer "jefferies_tube.add_maintenance_middleware" do |config|
+      if File.exists? "tmp/maintenance.txt"
+        require 'jefferies_tube/rack/maintenance'
+        config.middleware.use 'JefferiesTube::Rack::Maintenance'
+      end
+    end
+
+    private
```
Comment (boatrite) [inline lib/jefferies_tube/railtie.rb]: put new line after private

File: lib/jefferies_tube/railtie.rb [hunk 7]
```diff
@@ -18,7 +18,22 @@ class Railtie < ::Rails::Railtie
...
+    def initialize_404_catchall
+
```
Comment (boatrite) [inline lib/jefferies_tube/railtie.rb]: Probably don't need the new lines in an empty method.

**Comment by boatrite** [comment]
Small stuff and :shipit:

File: lib/jefferies_tube/rack/maintenance.rb (hunk 2 above)
Comment (samsonasu) [inline lib/jefferies_tube/rack/maintenance.rb]: derp

File: lib/jefferies_tube/railtie.rb (hunk 7 above)
Comment (samsonasu) [inline lib/jefferies_tube/railtie.rb]: probably dont need the method at all :)

Comment (samsonasu) [inline lib/jefferies_tube/railtie.rb]: probably dont need any of these methods :)

**Pull request merged**